### PR TITLE
[Consensus Voting] Vote collectors implementation

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/5787

### Context

`VoteCollectors` main purpose is to index collectors by view, create and provide access to them in concurrent safe way.
This PR implements core logic for `VoteCollectors` as well as unit tests. 
Same related refactoring and changes were done to `VoteAggregator` in scope of this PR.

### CODE REVIEW

File: consensus/hotstuff/voteaggregator/vote_aggregator_v2.go [hunk 1]
```diff
@@ -33,15 +37,27 @@ type VoteAggregatorV2 struct {
 var _ hotstuff.VoteAggregatorV2 = &VoteAggregatorV2{}
 
 // NewVoteAggregatorV2 creates an instance of vote aggregator
-func NewVoteAggregatorV2(notifier hotstuff.Consumer, highestPrunedView uint64, committee hotstuff.Committee, voteValidator hotstuff.Validator, signer hotstuff.SignerVerifier) *VoteAggregatorV2 {
-	return &VoteAggregatorV2{
+// NOTE: verifyingProcessorFactory has to be passed outside to abstract out knowledge about Consensus vs Collection
+// cluster running vote aggregation.
```
Comment (AlexHentschel) [inline consensus/hotstuff/voteaggregator/vote_aggregator_v2.go:41]:
```suggestion
// Note: verifyingProcessorFactory is injected. Thereby, the code is agnostic to the
// different voting formats of main Consensus vs Collector consensus.   
```

File: consensus/hotstuff/voteaggregator/vote_aggregator_v2.go [hunk 2]
```diff
@@ -33,15 +37,27 @@ type VoteAggregatorV2 struct {
...
 		voteValidator:     voteValidator,
 		signer:            signer,
 		unit:              engine.NewUnit(),
 	}
+
+	newCollectorFactoryMethod := func(view uint64) (hotstuff.VoteCollector, error) {
+		return votecollector.NewStateMachine(view, log, aggregator.workerPool, notifier, verifyingProcessorFactory), nil
+	}
+
+	aggregator.collectors = NewVoteCollectors(highestPrunedView, newCollectorFactoryMethod)
```
Comment (AlexHentschel) [inline consensus/hotstuff/voteaggregator/vote_aggregator_v2.go:59]: I would expect that it is easier to inject the implementation of `VoteCollectors`. Thereby, we can mock `VoteCollectors` for testing purposes.

**Review by AlexHentschel — APPROVED** [review]
Clean code. Great tests.

File: consensus/hotstuff/voteaggregator/vote_aggregator_v2.go (hunk 2 above)
Comment (durkmurder) [reply consensus/hotstuff/voteaggregator/vote_aggregator_v2.go:59]: Let's address this in scope of vote aggregator implementation. I don't want to add/remove too much of fields at this PR.

File: consensus/hotstuff/voteaggregator/vote_collectors.go [hunk 3]
```diff
@@ -1,31 +1,98 @@
...
+// This structure is concurrently safe.
+type VoteCollectors struct {
+	lock            sync.RWMutex
+	lowestLevel     uint64                            // lowest view that we have pruned up to
+	collectors      map[uint64]hotstuff.VoteCollector // view -> VoteCollector
+	createCollector NewCollectorFactoryMethod         // factory method for creating collectors
 }
 
-var _ hotstuff.VoteCollectors = &voteCollectors{}
+var _ hotstuff.VoteCollectors = &VoteCollectors{}
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/voteaggregator/vote_collectors.go:25]: I think if we move it to tests then we lose compilation error when we actually try to compile module without tests. Maybe I am getting something wrong?

File: consensus/hotstuff/votecollector/statemachine.go [hunk 4]
```diff
@@ -32,6 +32,8 @@ type VoteCollector struct {
 	votesProcessor atomic.Value
 }
 
+var _ hotstuff.VoteCollector = &VoteCollector{}
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/votecollector/statemachine.go:35]: True, but will it work if we try to build the module without tests?

**Review by zhangchiqing — APPROVED** [review]
Nice work 👍

**Pull request merged**